add review branch (#1)
Pearlicia/port-task: README.md

Reviewer: I don't see any frontend setup in port so how do we know how this automation is triggered

# YouTube Playlist & Videos Catalog in Port

> ## Introduction

Reviewer: change this to overview. Try following port's guide standards

Author: Done

> In this guide, I will demonstrate how Port can efficiently manage and visualize data from a YouTube playlist, helping you get insights at a glance. Port is a robust platform for modeling and cataloging diverse data types, and this guide will showcase its capabilities for managing video data.

Reviewer: remove the description about port. Focus the introduction on the value the guide brings

Author: Done

> ## Why Use Port?

Reviewer: This content is irrelevant to guide. we don't want to advertise port. we want to be more direct and straight forward on what the guide is doing

Author: Deleted

     4. On the new project dashboard, click Explore & Enable APIs.
     After creating a project, you’ll be taken to a brand new dashboard with different cards: Project Info, Resources, and so on.

     Navigate to the Getting Started card and click on Explore & Enable APIs.

     Alternatively, on the sidebar, navigate to APIs & Services > Library.

     5. In the library, navigate to YouTube Data API v3 under YouTube APIs.
     Once you’ve reached the library, you’ll see a page divided into sections. Navigate to the section titled YouTube APIs and select YouTube Data API v3.

     Alternatively, you can search for the YouTube Data API on the search bar.

     6. Enable the API.
     After you arrive at the YouTube Data API page, click a blue button with the word Enable.

     7. Create a credential.
     After clicking Enable, you’ll be taken to an overview page.

     Click on Credentials, Then Create Credentials. Select API Key

>     8. A screen will appear with the API key. Copy it. It will be also saved in your credentials for easy retrieval.

Reviewer: Simplify the prerequisite and make it just about the things you need for this guide. the prerequisites should not be a manual for actions. eg
```mrkdown
## Prerequisites

This guide assumes the following:
- You have a Port account and have completed the [onboarding process](https://docs.getport.io/quickstart).
- You have [installed and set up Port's New Relic integration](https://docs.getport.io/build-your-software-catalog/sync-data-to-catalog/apm-alerting/newrelic).
- You have the `New Relic Service Level` blueprint configured in your Port environment.
```

Author: Done

## Step 1: Model Data in Port

> 

Reviewer: This content speaks generally on the data you have to model, so the naming is wrong. 
Maybe you can use Data Model and explain that this section is about modeling the data and under data model create blueprints. would be the first step. and avoid the numbering in the headers

The first step is to define a model for the YouTube playlist and videos in Port. This involves creating **blueprints** for the playlist and each video, defining their properties, and establishing relationships.

> ### 1.1. Create Blueprints

Reviewer: This can be ```### Create blueprints```

- **Playlist Blueprint**: This represents the playlist as a whole.
  - **Properties**:

>     - **Playlist ID** (String)

Reviewer: this is not needed. the blueprint has json data in it. make this the json data they need. so they can just copy and paste.

    - **Title** (String)
    - **Description** (String)
    - **Published At** (Date time)
    - **Channel Id** (String)
    - **Channel Title** (String)
    
    ![Playlist-blueprint](./assets/playlist-blueprint.png)
    

> - **Video Blueprint**: Each video is an item within the playlist.

Reviewer: same goes here

  - **Properties**:
    - **Video ID** (String)
    - **Title** (String)
    - **Description** (String)
    - **Published At** (Date time)
    - **Duration** (String)
    - **View Count** (Number)
    - **Channel Id** (String)
    - **Channel Title** (String)
    - **Number of Comments** (Number)
    - **Number of Likes** (Number)
    - **Category Id** (String)
    - **Caption** (Number)
   
    
    ![Video-blueprint](./assets/video-playlist.png)

**Example JSON for Playlist Blueprint**:
```json

> {

Reviewer: Not example json, it should be `the` JSON. and have it in a collapsible dropdown

Author: Done

  "blueprint": "Playlist",
  "properties": {
    "Playlist Id": "string",
    "Title": "string",
    "Description": "string",
    "Published At": "datetime",
    "Channel Id": "string",
    "Channel Title": "string"
  }
}
```

**Example JSON for Video Blueprint**:
```json
{
  "blueprint": "Video",
  "properties": {
    "Video Id": "string",
    "Title": "string",
    "Description": "string",
    "Published At": "datetime",
    "Duration": "string",
    "View": "number",
    "Channel Id": "string",
    "Channel Title": "string",
    "Number of Comments": "number",
    "Number of Likes": "number",
    "Category Id": "string",
    "Caption": "string"

  },
  "relationships": {
    "belongs_to": "Playlist"
  }
}
```

### 1.2. Define Relationships
- Link each video to the playlist using a *one-to-many relationship*. This means one playlist blueprint can contain multiple video blueprints.

> Click on Video blueprint, then click on New relation

Reviewer: just have this in the blueprint, don't talk about it here

Author: Done

  - On Title, enter any title of your choice
  - On Description, Enter any description
  - On Related to, select Playlist
  - On Limit, select 1 entity
  - On Required, select False
  Then click Create

  ![Create-relation](./assets/create-relation.png)

  ![Relationship](./assets/relation-between-video-n-playlist.png)

## Step 2: Set Up GitHub Workflow

> 

Reviewer: minimal information. the header will be irrelevant

Author: Removed

This step involves creating a workflow to fetch data from YouTube and ingest it into Port.

### 2.1 Fetch Data from YouTube

> To fetch data, we’ll need to use the YouTube Data API. Here’s how to set up a basic request:

Reviewer: all this can be added as comments to your yaml file. They don't need this information here

Author: Fixed

1. **Get API Key**: Register for an API key on the [Google Developer Console](https://console.developers.google.com/). The steps to achieve that is written above on the prerequisites
2. **Fetch Playlist Data**:
   - Use the endpoint `https://www.googleapis.com/youtube/v3/playlists`.
   - Set parameters: `id` (playlist ID), `part` (snippet, contentDetails).

> **Example API Call**:

Reviewer: same here

Author: Removed

```bash
curl -X GET "https://www.googleapis.com/youtube/v3/playlists?id=YOUR_PLAYLIST_ID&part=snippet,contentDetails&key=YOUR_API_KEY"
```

3. **Fetch Video Data**:
   - Use the endpoint `https://www.googleapis.com/youtube/v3/playlistItems`.
   - Set parameters: `playlistId` (playlist ID), `part` (snippet, contentDetails).

**Example API Call**:
```bash
curl -X GET "https://www.googleapis.com/youtube/v3/playlistItems?playlistId=YOUR_PLAYLIST_ID&part=snippet,contentDetails&key=YOUR_API_KEY"
```

> ### 2.2 Ingest Data into Port

Reviewer: note that we want to do more of the action than the explanation

Author: Ok

Once the data is fetched, we'll use Port’s API to upload it.

1. **Format Data in JSON**: Ensure the data matches the blueprint properties defined earlier.
2. **Automate with GitHub Actions**:
   - Write a GitHub Action that runs the fetch and ingest commands.

>    Save the following in your GitHub secrets

Reviewer: you can explain this in an admonition. and skip this content. it keeps the guide simple and minimalistic

Author: Fixed

              env:
                ACCESS_TOKEN: ${{ env.access_token }}
                BLUEPRINT_ID: ${{ secrets.PORT_BLUEPRINT_ID }}
              run: |
                # Parse data.json and extract required fields for each video
                cat data.json | jq -c '.items[] | {
                  identifier: .contentDetails.videoId,
                  title: .snippet.title,
                  properties: {
                    id: .contentDetails.videoId,
                    title: .snippet.title,
                    description: .snippet.description,
                    channel_id: .snippet.channelId,
                    channel_title: .snippet.channelTitle,
                    published_at: .contentDetails.videoPublishedAt
                  }
                }' > processed_data.json

                # Loop through each video entry in processed_data.json and ingest to Port
                while IFS= read -r video; do
                  curl --location --request POST "https://api.getport.io/v1/blueprints/${BLUEPRINT_ID}/entities?upsert=true" \
                    --header "Authorization: Bearer $ACCESS_TOKEN" \
                    --header "Content-Type: application/json" \

>                     --data-raw "$video"

Reviewer: this should be in a detailed and summary tag, to make it collapsible

Author: Done

                done < processed_data.json

     ```

     ![Playlist-catalog](./assets/playlist-catalog.png)

     ![Video-entity](./assets/video-entity.png)

## Step 3: Visualize the Data in Port

> Port offers tools to visualize data, allowing you to track key metrics like view counts, publishing trends, and playlist growth. Here are two examples of valuable visualizations:

Reviewer: it should be something like this
`Track your data and visualize your YouTube data in port by ... ` something like that. this is a general input on ports visualization

Author: Done

# YouTube Playlist & Videos Catalog in Port

> In this guide, we will demonstrate how to efficiently manage and visualize data from a YouTube playlist, providing quick insights into your video content. By following this guide, you'll learn how to automate data ingestion, processing, and upsertion to streamline your workflow for managing video data.

Author: Done

## Prerequisites
1. GitHub Account 
2. You have a Port account and have completed the [onboarding process](https://docs.getport.io/quickstart). 
3. Youtube Playlist.

> 4. YouTube data API V3 Key.

Reviewer: Remove this and talk to them about generating a YouTube API key when it's time to use it, you can attach a link to direct them to how to create a YouTube API key

## Data Modeling
The first step is to define a model for the YouTube playlist and videos in Port. This involves creating **blueprints** for the playlist and each video, defining their properties, and establishing relationships.

### Create Blueprints

> - **Playlist Blueprint**: This represents the playlist as a whole.

Reviewer: ```suggestion
 Add the JSON snippet below to create the blueprint
```

> - **Playlist Blueprint**: This represents the playlist as a whole.

Reviewer: But before this tell them where they have to go to be able to add this blueprint. 
We can't assume they know where to go

Author: Done

<details>
  <summary><strong>JSON for Playlist Blueprint</strong></summary>
  
  ```json

   {
     "identifier": "playlist",
     "description": "A YouTube playlist",
     "title": "Playlist",
     "icon": "Youtrack",
     "schema": {
       "properties": {
         "id": {
           "type": "string",

>            "title": "Id",

Reviewer: I'd recommend you add playlist to it to make it more specific

           "description": "Playlist id"
         },
         "title": {
           "type": "string",

>            "title": "Title",

Reviewer: this will conflict with the title you already have up there, add more adjective to make it specific

Author: Fixed

           "description": "Title "
         },
         "description": {
           "type": "string",
           "title": "Description",
           "description": "Description"

>          },

Reviewer: this as well. they are all meta properties. you can read more about ports meta properties to know which of these names are taken already

Author: Fixed

         "channel_id": {

>            "type": "string",

Reviewer: this is good because it's not a meta property let the ones above be like this

           "title": "Channel Id",
           "description": "Channel id"
         },
         "channel_title": {
           "type": "string",
           "title": "Channel Title"
         },
         "published_at": {
           "type": "string",
           "title": "Published At",
           "description": "Published date",
           "format": "date-time"
         },
       },
       "required": []
     },
     "mirrorProperties": {},
     "calculationProperties": {},
     "aggregationProperties": {},
     "relations": {}
   }

  ```
</details>

> ![Playlist-blueprint](./assets/playlist-blueprint.png)

Reviewer: use an img tag here and point load the image with the source like 
   <img src="/image_path/image_name.png" /> and add some borders

<details>

>   <summary><strong>JSON for Video Blueprint</strong></summary>

Reviewer: don't use strong, use <b></b>

Author: Done

>   ```json

Reviewer: ```suggestion
  ```json showLineNumber
```

     {
        "identifier": "video",
        "description": "YouTube videos in a a playlist",
        "title": "Video",
        "icon": "Youtrack",
        "schema": {
          "properties": {
            "id": {
              "icon": "Youtrack",
              "type": "string",
              "description": "Video d",
              "title": "Id"
            },
            "description": {
              "type": "string",

>               "title": "Description",

Reviewer: Same as the comment above

Author: Done

              "title": "Channel Title",
              "description": "Playlist channel title "
            },
            "category_id": {
              "type": "string",
              "title": "Category Id"
            },
            "duration": {
              "type": "string",
              "title": "Duration"
            },
            "caption": {
              "type": "string",
              "title": "Caption"
            },
            "view_count": {
              "type": "string",
              "title": "view count"
            },
            "like_count": {
              "type": "string",
              "title": "like count"
            },
            "comment_count": {
              "type": "string",
              "title": "comment count"
            }
          },
          "required": []
        },
        "mirrorProperties": {},
        "calculationProperties": {},
        "aggregationProperties": {},
        "relations": {
          "playlist": {
            "title": "Playlist",
            "description": "Playlist",
            "target": "playlist",
            "required": false,
            "many": false
          }
        }
      }
  
  ```
</details>

> ![Video-blueprint](./assets/video-playlist.png)

Reviewer: same as the comment with the image

Author: Done

> ## Workflow to Ingest Data into Port

Reviewer: Find a better title. 
Ingest workflow

> :::info Save the following in your GitHub secrets

Reviewer: show them how to do that how to save you can direct them to a GitHub add secret link

:::

  - PORT_CLIENT_ID
  - PORT_CLIENT_SECRET
  - PORT_BLUEPRINT_ID
  - YOUTUBE_API_KEY
  - YOUTUBE_PLAYLIST_ID

<details>

>   <summary><strong>Workflow</strong></summary>

Reviewer: ```suggestion
  <summary><b>Workflow</b></summary>
```

Reviewer: Also use something Workflow (click to expand) for the summary heading

          run: |
            while IFS= read -r video; do
              curl --location --request POST "https://api.getport.io/v1/blueprints/${BLUEPRINT_ID}/entities?upsert=true" \
                --header "Authorization: Bearer $ACCESS_TOKEN" \
                --header "Content-Type: application/json" \
                --data-raw "$video"
            done < processed_data.json

    report-status:
      runs-on: ubuntu-latest
      needs: upsert-to-port
      env:
        ACCESS_TOKEN: ${{ env.access_token }}
      steps:
        - name: Report Workflow Status to Port
          run: |
            status="success"
            if [ "${{ needs.upsert-to-port.result }}" != "success" ]; then
              status="failure"
            fi

            # Report the status back to Port
            curl --location --request POST "https://api.getport.io/v1/status" \
              --header "Authorization: Bearer $ACCESS_TOKEN" \
              --header "Content-Type: application/json" \
              --data-raw '{
                "status": "'"$status"'",
                "details": "Workflow completed with status: '"$status"'"
              }'
  ```
</details>

   

> ![Playlist-catalog](./assets/playlist-catalog.png)

Reviewer: You know what to do here

![Video-entity](./assets/video-entity.png)

## Visualize the Data in Port

Track your data and visualize your YouTube data in port by key metrics like view counts, publishing trends, and playlist growth. Here are two examples of valuable visualizations:

> ### Visualization 1: Videos count

Reviewer: Be more descriptive about how they can set up the widget to look like how you set up yours in the picture. 
Refer to dora metrics visualization for more context on this https://docs.getport.io/guides/all/setup-dora-metrics/#visualization

- **Objective**: Show the number of videos published over time.
- **Usage**: Identify the number of videos added to the playlist.

![Videos-count](./assets/dashboard-vid-count.png)

### Visualization 2: View Count Distribution
- **Objective**: Display the view count of each video to identify popular content.
- **Usage**: Understand which topics or types of content resonate most with viewers.

> 

Reviewer: I don't see how the automation is triggered. where's the frontend

> # YouTube Playlist & Videos Catalog in Port

Reviewer: ```suggestion
# Ingesting YouTube playlist and videos in Port
```

In this guide, we will demonstrate how to efficiently manage and visualize data from a YouTube playlist, providing quick insights into your video content. 
By following this guide, you'll learn how to ingest your YouTube data into Port

## Prerequisites
 - GitHub Account 
 - You have a Port account and have completed the [onboarding process](https://docs.getport.io/quickstart) 
 - Youtube Playlist
 - Youtube data API key. [Get your Youtube API key](https://blog.hubspot.com/website/how-to-get-youtube-api-key)

> ## Data Modeling

Reviewer: ```suggestion
## Data modeling
```

The first step is to define a model for the YouTube playlist and videos in Port. This involves creating **blueprints** for the playlist and each video, defining their properties, and establishing relationships.

### Create Blueprints

>   - Log in to Port and click on `Builder`

Reviewer: add a link to the builder page here

Author: Don't understand

Author: Done

  - Then click on the `+ Blueprint` button
  - Click on `Edit JSON`

> Copy and paste the JSON snippet below to create the blueprint

Reviewer: ```suggestion
Add the JSON snippet below to create the playlist blueprint
```

<details>

>   <summary><b>Playlist Blueprint JSON (Click to expand)</b></summary>

Reviewer: ```suggestion
  <summary><b>Playlist blueprint (click to expand)</b></summary>
```

        },
        "channel_title": {
          "type": "string",
          "title": "Channel Title"
        },
        "published_at": {
          "type": "string",
          "title": "Published At",
          "description": "Published date",
          "format": "date-time"
        },
        "view_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "View count"
        },
        "like_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Like count"
        },
        "comment_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Comment count"
        },
        "duration": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Duration"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "aggregationProperties": {},
    "relations": {
      "video": {
        "title": "Video",
        "description": "Has videos",
        "target": "playlist",
        "required": false,
        "many": false
      }
    }
  }       
  ```
</details>

> <img src="./assets/new-p-blueprint.png" alt="Playlist-blueprint" style="border: 2px solid black; border-radius: 5px;">

Reviewer: ```suggestion
<img src="./assets/new-p-blueprint.png" alt="Playlist-blueprint" style="border: 1px">
```

> 

Reviewer: ```suggestion

Add the snippet below to create the video blueprint:

```

<details>

>   <summary><b> Video Blueprint JSON (Click to expand)</b></summary>

Reviewer: ```suggestion
  <summary><b> Video blueprint (click to expand)</b></summary>
```

        },
        "category_id": {
          "type": "string",
          "title": "Category Id"
        },
        "duration": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Duration"
        },
        "caption": {
          "type": "string",
          "title": "Caption"
        },
        "view_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "View count"
        },
        "like_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Like count"
        },
        "comment_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Comment count"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "aggregationProperties": {},
    "relations": {
      "playlist": {
        "title": "Playlist",
        "description": "Playlist",
        "target": "playlist",
        "required": false,
        "many": false
      }
    }
  }
  ```
</details>

<img src="./assets/new-v-blueprint.png" alt="Video-blueprint" style="border: 2px solid black; border-radius: 5px;">

> ## Ingest Data

Reviewer: ```suggestion
## Ingest Youtube data
```

To fetch and ingest YouTube playlist data into Port, we'll use Port's REST API integrated within a GitHub Actions pipeline.

> Create a yaml file in your Github repository under .github/workflows 

Reviewer: ```suggestion
Follow the steps below:

1.  Create a yaml file in your Github repository under .github/workflows 
```

> Copy and paste the workflow below

Reviewer: ```suggestion
2. Copy and paste the workflow below
```

:::info 
Save the following in your GitHub secrets:
- **PORT_CLIENT_ID**: Your Port client ID
- **PORT_CLIENT_SECRET**: Your Port client secret
- **YOUTUBE_API_KEY**: Your YouTube Data API key
- **PERSONAL_ACCESS_TOKEN**: Generate a Personal Access Token (PAT) with the following scopes:
     - **repo** (for repository access)
     - **workflow** (to trigger workflows)
:::

> :::info

Reviewer: too many admonitions put all of them into one and separate them by lines

Author: Done

Change YOUR_PLAYLIST_ID to your playlist ID 
:::

> 

Reviewer: this is confusing

Reviewer 2: yes a better approach will be to actually make the self service action( SSA) dynamic so that the playlist ID is passed from Port. This way, you can use the same SSA to create multiple playlist, without having to always come to Github workflow and modify the code.
So when the SSA is executed, ask the user for the playlist ID, and use it to fetch the data from youtbue api

Author: Fixed

                nextPageToken=$(echo "$response" | jq -r '.nextPageToken')
                [ "$nextPageToken" == "null" ] && break
            done

    report-status:
      runs-on: ubuntu-latest
      needs: upsert-playlist-data
      steps:
        - name: Report Status to Port
          env:
            ACCESS_TOKEN: ${{ needs.get-access-token.outputs.access_token }}
            PERSONAL_ACCESS_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
          run: |
            status="Playlist ingestion completed successfully"
            curl --location --request POST "https://api.getport.io/v1/status" \
            --header "Authorization: Bearer $ACCESS_TOKEN" \
            --header "Content-Type: application/json" \
            --data-raw "{
                \"status\": \"$status\",
                \"description\": \"YouTube playlist ingestion job completed.\"
            }"
            echo "Reported status back to Port."
     
  ```
</details>

## Create Self Service 
We'll create a self service to trigger the workflow to run from Port.

### Create Action
We'll create an action that triggers the playlist workflow file.

**To create the action:**

> :::info 

Reviewer: Add titles to admonitions so they describe what the admonition is for

Author: Done

<img src="./assets/self-service-success.png" alt="service-success" style="border: 2px solid black; border-radius: 5px;">

**If Self service action is `Successful`**
Navigate to Catalog and click on Playlist service to view your playlist video data

<img src="./assets/new-p-catalog.png" alt="Playlist-catalog" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/new-p-data-page.png" alt="Video-entity" style="border: 2px solid black; border-radius: 5px;">

## Visualization

Let's visualize our YouTube data in Port by key metrics like view counts, and number of videos in the playlist.

### Setup
   - Click on catalog.
   - Click on the `+ New` button in the left sidebar.
   - Select `New dashboard`.
   - On `Title` type **Playlist Videos**
   - On `Description` type **Playlist Videos Dashboard**
   - choose an icon if desired, or leave as default
   - click Create.

<img src="./assets/Create-dashboard.jpg" alt="create-dashboard" style="border: 2px solid black; border-radius: 5px;">

**This will create a new empty dashboard.**

### Adding widgets

#### Videos Count
Let's display the number of videos in the playlist using a number chart

>    - Click **+ Widget** and select Number Chart.

Reviewer: Steps should be numbered

Author: Done

# Ingesting YouTube playlist and videos in Port

In this guide, we will demonstrate how to efficiently manage and visualize data from a YouTube playlist, providing quick insights into your video content. 
By following this guide, you'll learn how to ingest your YouTube data into Port

## Prerequisites
 - GitHub Account 
 - You have a Port account and have completed the [onboarding process](https://docs.getport.io/quickstart) 
 - Youtube Playlist
 - Youtube data API key. [Get your Youtube API key](https://blog.hubspot.com/website/how-to-get-youtube-api-key)

> ## Data modeling
> The first step is to define a model for the YouTube playlist and videos in Port. This involves creating **blueprints** for the playlist and each video, defining their properties, and establishing relationships.

Reviewer 2: ```suggestion
## Data modeling
The first step is to define a model for the YouTube playlist and videos in Port. This involves creating **blueprints** for the playlist and video, defining their properties, and establishing relationships.
```

Author: Fixed

          "title": "Channel Id",
          "description": "Channel id"
        },
        "channel_title": {
          "type": "string",
          "title": "Channel Title"
        },
        "published_at": {
          "type": "string",
          "title": "Published At",
          "description": "Published date",
          "format": "date-time"
        },
        "view_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "View count"
        },
        "like_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Like count"
        },
        "comment_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Comment count"
        },
        "duration": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Duration"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "aggregationProperties": {},
    "relations": {
      "video": {
        "title": "Video",
        "description": "Has videos",
        "target": "playlist",
        "required": false,

>         "many": false

Reviewer 2: I want you to rethink this relationship. A playlist has a list of videos while a video may be part of a playlist. To define this relationship. I will rather create a playlist relation on the video blueprint. signalling that this video is a member of the referenced playlist. what you have done looks like the opposite

          "type": "string",
          "title": "Category Id"
        },
        "duration": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Duration"
        },
        "caption": {
          "type": "string",
          "title": "Caption"
        },
        "view_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "View count"
        },
        "like_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Like count"
        },
        "comment_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Comment count"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "aggregationProperties": {},
    "relations": {
      "playlist": {
        "title": "Playlist",
        "description": "Playlist",
        "target": "playlist",
        "required": false,
        "many": false
      }
    }
  }
  ```
</details>

<img src="./assets/new-v-blueprint.png" alt="Video-blueprint" style="border: 2px solid black; border-radius: 5px;">

## Ingest Youtube data

> To fetch and ingest YouTube playlist data into Port, we'll use Port's REST API integrated within a GitHub Actions pipeline.

Reviewer 2: remove the pipeline at the end of the sentence

Author: Done

Follow the steps below:

1.  Create a yaml file in your Github repository under .github/workflows 
2. Copy and paste the workflow below

:::info Save credentials

Save the following in your GitHub secrets:
- **PORT_CLIENT_ID**: Your Port client ID
- **PORT_CLIENT_SECRET**: Your Port client secret
- **YOUTUBE_API_KEY**: Your YouTube Data API key
- **PERSONAL_ACCESS_TOKEN**: Generate a Personal Access Token (PAT) with the following scopes:
     - **repo** (for repository access)
     - **workflow** (to trigger workflows)
:::

:::info Add YouTube playlist ID
- Change YOUR_PLAYLIST_ID to your playlist ID 
:::

:::info Add blueprint ID

> - Change blueprint_id to your blueprint ID if it's not **playlist** 

Reviewer 2: this information can be received in the port_context github variable so you don't have to make the user provide it

Author: Fixed

          run: sudo apt-get install -y jq

        - name: Get Access Token
          id: get-token
          env:
            PORT_CLIENT_ID: ${{ secrets.PORT_CLIENT_ID }}
            PORT_CLIENT_SECRET: ${{ secrets.PORT_CLIENT_SECRET }}
          run: |
            # Request access token
            access_token=$(curl --location --request POST 'https://api.getport.io/v1/auth/access_token' \
            --header 'Content-Type: application/json' \
            --data-raw "{
                \"clientId\": \"$PORT_CLIENT_ID\",
                \"clientSecret\": \"$PORT_CLIENT_SECRET\"
            }" | jq -r '.accessToken')
            echo "::set-output name=access_token::$access_token"

    upsert-playlist-data:
      runs-on: ubuntu-latest
      needs: get-access-token
      env:
        YOUTUBE_API_KEY: ${{ secrets.YOUTUBE_API_KEY }}
        ACCESS_TOKEN: ${{ needs.get-access-token.outputs.access_token }}
      steps:
        - name: Install jq
          run: sudo apt-get install -y jq

        - name: Fetch and Upsert YouTube Playlist Data to Port
          run: |
            # Set the playlist ID and blueprint ID
            PLAYLIST_ID="YOUR_PLAYLIST_ID"

>             blueprint_id="playlist"

Reviewer 2: these information should be received by the github worklow under the port_context variable. check here. the workflow can receive variables from Port

https://docs.getport.io/guides/all/change-status-and-assignee-of-jira-ticket#github-workflow

Author: Done

                done

                # Check if there is another page of results
                nextPageToken=$(echo "$response" | jq -r '.nextPageToken')
                [ "$nextPageToken" == "null" ] && break
            done

    report-status:
      runs-on: ubuntu-latest
      needs: upsert-playlist-data
      steps:
        - name: Report Status to Port
          env:
            ACCESS_TOKEN: ${{ needs.get-access-token.outputs.access_token }}
            PERSONAL_ACCESS_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
          run: |
            status="Playlist ingestion completed successfully"
            curl --location --request POST "https://api.getport.io/v1/status" \
            --header "Authorization: Bearer $ACCESS_TOKEN" \
            --header "Content-Type: application/json" \
            --data-raw "{
                \"status\": \"$status\",
                \"description\": \"YouTube playlist ingestion job completed.\"
            }"
            echo "Reported status back to Port."
     
  ```
</details>

## Create Self Service 
We'll create a self service to trigger the workflow to run from Port.

> ### Create Action

Reviewer 2: delete this header

We'll create a self service to trigger the workflow to run from Port. We will start by creating an action that triggers the playlist Github workflow

Author: Done

We'll create an action that triggers the playlist workflow file.

**To create the action:**

:::info Edit credentials
Change 
- **<GITHUB_ORG_NAME>** and
- **<GITHUB_REPO_NAME>** with your organization and repository name.
:::

1. Log in to Port and click on the **Self-service** tab.
2. Click on **+ Action** button.
3. Click on **Edit JSON**.
4. Copy and paste the JSON below
5. Click Save

<details>
    <summary><b>Action JSON (Click to expand)</b></summary>

   ```json showLineNumber
   {
     "identifier": "trigger_github_workflow",
     "title": "Trigger GitHub Workflow",
     "icon": "Github",
     "description": "Triggers a GitHub workflow in the repository",
     "trigger": {
       "type": "self-service",
       "operation": "CREATE",
       "userInputs": {
         "properties": {
           "workflow_name": {
             "icon": "DefaultProperty",
             "title": "Playlist workflow action",
             "type": "string"

>            }

Reviewer 2: the user does not need to provide the workflow name. this should be already configured in the SSA backend

Author: Done

         },
         "required": ["workflow_name"],
         "order": ["workflow_name"]
       },
       "blueprintIdentifier": "playlist"
     },
     "invocationMethod": {
       "type": "GITHUB",
       "org": "<GITHUB_ORG_NAME>",
       "repo": "<GITHUB_REPO_NAME>",

>        "workflow": "{{ .inputs.\"workflow_name\" }}",

Reviewer 2: the workflow name should be static. so make the user update this field with the name of the workflow such as create_playlist_workflow.yml

Author: Done

       "reportWorkflowStatus": true
     },
     "requiredApproval": false
   }
   ```
</details>

- **`workflow_name`** is the name of the workflow file you want to trigger (e.g., `run-workflow.yml`).
- **`reportWorkflowStatus`** allows Port to report the workflow status back to the Port dashboard.

### **Trigger the Action**

To execute the action:
  1. Open the **Self-service** tab in Port.
  2. Select **Trigger GitHub Workflow**.
  3. Enter the name of the workflow file (e.g., `run-workflow.yml`) as the `workflow_name`.
  4. Submit the form, and the GitHub workflow will run automatically.

You can view the status of the action by clicking on the progress icon at the top right corner on the page
As shown on the images below 

<img src="./assets/self-service.png" alt="Self-service" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/self-service-success.png" alt="service-success" style="border: 2px solid black; border-radius: 5px;">

**If Self service action is `Successful`**

> Navigate to Catalog and click on Playlist service to view your playlist video data

Reviewer 2: remove the service

`Navigate to Catalog and click on Playlist to view your playlist video data
`

Author: Done

<img src="./assets/new-p-catalog.png" alt="Playlist-catalog" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/new-p-data-page.png" alt="Video-entity" style="border: 2px solid black; border-radius: 5px;">

## Visualization

Let's visualize our YouTube data in Port by key metrics like view counts, and number of videos in the playlist.

### Setup

>    1. Click on catalog.

Reviewer 2: `Catalog` in Port

Author: Done

   2. Click on the `+ New` button in the left sidebar.
   3. Select `New dashboard`.
   4. On `Title` type **Playlist Videos**

>    5. On `Description` type **Playlist Videos Dashboard**

Reviewer 2: instead of `type` use `enter`

Author: Fixed

   6. choose an icon if desired, or leave as default

>    7. click Create.

Reviewer 2: please pay attention to your formatting. you should start your sentences with capital characters

Author: Fixed

>    7. click Create.

Reviewer 2: Click on the `Create` button to create your dashboard

Author: Fixed

<img src="./assets/Create-dashboard.jpg" alt="create-dashboard" style="border: 2px solid black; border-radius: 5px;">

**This will create a new empty dashboard.**

> ### Adding widgets

Reviewer 2: what are widgets. you could explain that we will then create 2 widgets to display the number of likes. comments or views

Author: Done

#### Videos Count
Let's display the number of videos in the playlist using a number chart

> 

Reviewer 2: you should add a text that create a widget by providing the following information:

Author: I don't understand

Author: I don't understand

  
  ```json showLineNumber
  {
    "identifier": "playlist",
    "description": "A YouTube playlist",
    "title": "Playlist",
    "icon": "Youtrack",
    "schema": {
      "properties": {
        "description": {
          "type": "string",
          "title": "Playlist data",
          "description": "This is playlist blueprint description"
        },
        "channel_id": {
          "type": "string",
          "title": "Channel Id",
          "description": "Channel id"
        },
        "channel_title": {
          "type": "string",
          "title": "Channel Title"
        },
        "published_at": {
          "type": "string",
          "title": "Published At",
          "description": "Published date",
          "format": "date-time"
        },
        "view_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "View count"
        },
        "like_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Like count"
        },
        "comment_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Comment count"

>         },

Reviewer 2: these all should be number data type instead of string. in your jq (from the github workflow), you can to the `.video.like_count | tonumber` to convert the string to number so it can be inserted to Port

Author: This works.

        "duration": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Duration"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "aggregationProperties": {},
    "relations": {
      "video": {
        "title": "Video",
        "description": "Has videos",
        "target": "playlist",
        "required": false,
        "many": true

>       }

Reviewer 2: delete this relation since you already have a relation between playlist and video in the video blueprint. adding it here creates a cyclic relation

Reviewer 2: when done, update the image

          "description": "Video blueprint description "
        },
        "published_at": {
          "icon": "Youtrack",
          "type": "string",
          "title": "Published At",
          "description": "Date published ",
          "format": "date-time"
        },
        "channel_id": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Channel ID",
          "description": "Playlist channel id"
        },
        "channel_title": {
          "type": "string",
          "title": "Channel Title",
          "description": "Playlist channel title "
        },
        "category_id": {
          "type": "string",
          "title": "Category Id"
        },
        "duration": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Duration"
        },
        "caption": {
          "type": "string",
          "title": "Caption"
        },
        "view_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "View count"
        },
        "like_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Like count"

>         },

Reviewer 2: these should be number data type instead of string

Author: Ok

        "comment_count": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Comment count"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "aggregationProperties": {},
    "relations": {
      "playlist": {
        "title": "Playlist",
        "description": "Playlist",
        "target": "playlist",
        "required": false,
        "many": false
      }
    }
  }
  ```
</details>

<img src="./assets/new-v-blueprint.png" alt="Video-blueprint" style="border: 2px solid black; border-radius: 5px;">

## Ingest Youtube data
To fetch and ingest YouTube playlist data into Port, we'll use Port's REST API integrated within a GitHub Actions.

Follow the steps below:

1. Create a yaml file in your Github repository under .github/workflows 
2. Copy and paste the workflow below

:::info Save credentials

> Save the following in your GitHub secrets:

Reviewer 2: rephrase to `Create the following GitHub secrets:`

Author: Done

- **PORT_CLIENT_ID**: Your Port client ID
- **PORT_CLIENT_SECRET**: Your Port client secret
- **YOUTUBE_API_KEY**: Your YouTube Data API key
- **PERSONAL_ACCESS_TOKEN**: Generate a Personal Access Token (PAT) with the following scopes:
     - **repo** (for repository access)

>      - **workflow** (to trigger workflows)

Reviewer 2: why do we need this?

Author: I don't understand the question

Author: Seems I don't have that anymore

:::

<details>
  <summary><b>Workflow yaml (Click to expand)</b></summary>

  ```yaml

>   name: YouTube Playlist to Port Ingestion

Reviewer 2: ```suggestion
  name: Ingest YouTube Playlist to Port
```

            PORT_CLIENT_SECRET: ${{ secrets.PORT_CLIENT_SECRET }}
          run: |
            # Request access token
            access_token=$(curl --location --request POST 'https://api.getport.io/v1/auth/access_token' \
            --header 'Content-Type: application/json' \
            --data-raw "{
                \"clientId\": \"$PORT_CLIENT_ID\",
                \"clientSecret\": \"$PORT_CLIENT_SECRET\"
            }" | jq -r '.accessToken')
            echo "::set-output name=access_token::$access_token"
  
    upsert-playlist-data:
      runs-on: ubuntu-latest
      needs: get-access-token
      env:
        YOUTUBE_API_KEY: ${{ secrets.YOUTUBE_API_KEY }}
        ACCESS_TOKEN: ${{ needs.get-access-token.outputs.access_token }}
      steps:
        - name: Install jq
          run: sudo apt-get install -y jq
  
        - name: Fetch and Upsert YouTube Playlist Data to Port
          run: |
            # Set the playlist ID
            PLAYLIST_ID= ${{ fromJson(inputs.port_context).playlist_id }}
  
            # Initialize nextPageToken
            nextPageToken=""
            while true; do
                # Fetch videos in the playlist (one page at a time)

>                 response=$(curl -s "https://www.googleapis.com/youtube/v3/playlistItems?part=snippet,contentDetails&playlistId=$PLAYLIST_ID&maxResults=50&pageToken=$nextPageToken&key=$YOUTUBE_API_KEY")

Reviewer 2: this job for upserting playlist data works very well. But earlier on, you made users create two blueprints: playlist and video. In this github workflow, I see you are inserting data only into one blueprint referenced as ${blueprint_id}. I assume that is video blueprint? Also, what about the playlist itself? I think you should fetch the playlist metadata and ingest it to Port as well.

Author: No, it's playlist blueprint

                                   '{
                                       identifier: $identifier,
                                       title: $title,
                                       properties: {
                                           id: $id,
                                           description: $description,
                                           title: $title,
                                           published_at: $published_at,
                                           channel_id: $channel_id,
                                           channel_title: $channel_title,
                                           duration: $duration,
                                           view_count: $view_count,
                                           like_count: $like_count,
                                           comment_count: $comment_count
                                       },
                                       relations: {
                                           playlist: "related_entity_identifier"
                                       }
                                   }')
  
                    # Ingest data into Port

>                     curl --location --request POST "https://api.getport.io/v1/blueprints/${blueprint_id}/entities?upsert=true" \

Reviewer 2: I see reference to ${blueprint_id} but I don't see where you have defined it

**To create the action:**

:::info Edit credentials

Replace
- **org**: With your GitHub organization name
- **repo**: With your repository name
- **workflow**: With your workflow name

:::

1. Log in to Port and click on the **Self-service** tab.
2. Click on **+ Action** button.
3. Click on **Edit JSON**.
4. Copy and paste the JSON below
5. Click Save

<details>
    <summary><b>Action JSON (Click to expand)</b></summary>

   ```json showLineNumber
  {
    "identifier": "playlist_service",
    "title": "Playlist service",
    "description": "This is the action to trigger playlist workflow",
    "trigger": {
      "type": "self-service",
      "operation": "CREATE",
      "userInputs": {
        "properties": {
          "playlist_id": {
            "icon": "DefaultProperty",
            "type": "string",
            "title": "Playlist Id",
            "description": "Youtube playlist id"
          }
        },
        "required": [
          "playlist_id"
        ],
        "order": [
          "playlist_id"
        ]
      },
      "blueprintIdentifier": "playlist"
    },
    "invocationMethod": {
      "type": "GITHUB",
      "org": "Pearlicia",
      "repo": "play-list-repo",

>       "workflow": "playlist.yml",

Reviewer 2: by convention, we just use placeholders such as <GITHUB-ORG> or <GITHUB-REPO-NAME> instead of passing your own details

Author: Ok

    },
    "requiredApproval": false
  }
   ```
</details>

- **`reportWorkflowStatus`** allows Port to report the workflow status back to the Port dashboard.

### **Trigger the Action**

To execute the action:
  1. Open the **Self-service** tab in Port.
  2. Select **Playlist service**.
  3. Enter your playlist ID (e.g., `PL-7fHWfrqQvEUgrFz5pE-NWd6-8ixbX8w`)
  4. Click **Execute**, and the GitHub workflow will run automatically.

You can view the status of the action by clicking on the progress icon, and then the service name at the top right corner on the page.
As shown on the images below 

<img src="./assets/new-ssa.png" alt="Self-service" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/new-ssa-stat.png" alt="service-success" style="border: 2px solid black; border-radius: 5px;">

**If Self service action is `Successful`**
Navigate to Catalog and click on Playlist to view your playlist video data

<img src="./assets/new-p-catalog.png" alt="Playlist-catalog" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/new-p-data-page.png" alt="Video-entity" style="border: 2px solid black; border-radius: 5px;">

## Visualization

> Let's visualize our YouTube data in Port by key metrics like view counts, and number of videos in the playlist.

Reviewer 2: ```suggestion
Let's visualize our YouTube data in Port by defining key metrics like view counts, and number of videos in the playlist.
```

# YouTube Playlist & Videos Catalog in Port

In this guide, we will demonstrate how to efficiently manage and visualize data from a YouTube playlist, providing quick insights into your video content. 
By following this guide, you'll learn how to ingest your YouTube data into Port

## Prerequisites
 - GitHub Account 
 - You have a Port account and have completed the [onboarding process](https://docs.getport.io/quickstart). 

>  - Youtube Playlist.

Author: Done

         },
         "channel_id": {
           "type": "string",
           "title": "Channel Id",
           "description": "Channel id"
         },
         "channel_title": {
           "type": "string",
           "title": "Channel Title"
         },
         "published_at": {
           "type": "string",
           "title": "Published At",
           "description": "Published date",
           "format": "date-time"
         },
         "view_count": {
           "type": "string",
           "title": "view count"
         },
         "like_count": {
           "type": "string",
           "title": "like count"
         },
         "comment_count": {
           "type": "string",
           "title": "comment count"
         },
         "duration": {
           "type": "string",
           "title": "duration"
         }
       },
       "required": []
     },
     "mirrorProperties": {},
     "calculationProperties": {},
     "aggregationProperties": {},
     "relations": {}
   }
        
  ```
</details>

<img src="./assets/playlist-blueprint.png" alt="Playlist-blueprint" style="border: 2px solid black; border-radius: 5px;">

<details>

>   <summary><b>JSON for Video Blueprint (Click to expand)</b></summary>

Reviewer: ```suggestion
  <summary><b> Video Blueprint (Click to expand)</b></summary>
```

          "type": "string",
          "title": "Category Id"
        },
        "duration": {
          "type": "string",
          "title": "Duration"
        },
        "caption": {
          "type": "string",
          "title": "Caption"
        },
        "view_count": {
          "type": "string",
          "title": "view count"
        },
        "like_count": {
          "type": "string",
          "title": "like count"
        },
        "comment_count": {
          "type": "string",
          "title": "comment count"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "aggregationProperties": {},
    "relations": {
      "playlist": {
        "title": "Playlist",
        "description": "Playlist",
        "target": "playlist",
        "required": false,
        "many": false
      }
    }
  }
  
  ```
</details>

<img src="./assets/video-playlist.png" alt="Video-blueprint" style="border: 2px solid black; border-radius: 5px;">

## Ingest Data

Create a yaml file in your Github repository under .github/workflows 
Copy and paste the workflow below

:::info Save the following in your GitHub secrets
:::

> 

Reviewer 3: Before the workflow YAML, add a brief overview of what the workflow will do.

Author: Done

  - **PORT_CLIENT_ID**
  - **PORT_CLIENT_SECRET**
  - **YOUTUBE_API_KEY**
  - **PERSONAL_ACCESS_TOKEN**: Generate a Personal Access Token (PAT) with the following scopes:
     - **repo** (for repository access)
     - **workflow** (to trigger workflows)

::info Change YOUR_PLAYLIST_ID to your playlist ID ::

[Get your Youtube API key](https://blog.hubspot.com/website/how-to-get-youtube-api-key)

<details>

>   <summary><b>WORKFLOW YAML (Click to expand)</b></summary>

Reviewer: ```suggestion
  <summary><b>Workflow yaml (Click to expand)</b></summary>
```

      steps:
        - name: Checkout code
          uses: actions/checkout@v2

        - name: Install jq
          run: sudo apt-get install -y jq

        - name: Get Access Token
          id: get-token
          env:
            PORT_CLIENT_ID: ${{ secrets.PORT_CLIENT_ID }}
            PORT_CLIENT_SECRET: ${{ secrets.PORT_CLIENT_SECRET }}
          run: |
            # Request access token
            access_token=$(curl --location --request POST 'https://api.getport.io/v1/auth/access_token' \
            --header 'Content-Type: application/json' \
            --data-raw "{
                \"clientId\": \"$PORT_CLIENT_ID\",
                \"clientSecret\": \"$PORT_CLIENT_SECRET\"
            }" | jq -r '.accessToken')
            echo "::set-output name=access_token::$access_token"

    upsert-playlist-data:
      runs-on: ubuntu-latest
      needs: get-access-token
      env:
        YOUTUBE_API_KEY: ${{ secrets.YOUTUBE_API_KEY }}
        ACCESS_TOKEN: ${{ needs.get-access-token.outputs.access_token }}
      steps:
        - name: Install jq
          run: sudo apt-get install -y jq

        - name: Fetch and Upsert YouTube Playlist Data to Port
          run: |

>             # Set the playlist ID and blueprint ID

Reviewer: You are asking for blueprint ID to be set here. without properly looking at this yaml one might miss this information. None the less playlist id can be retrieved from the port context.  if you insist on making them set it in the workflow which I would strongly discourage then leave a tip or info that they should set replay YOUR BLUEPRINT ID with their actual blueprint Id

      steps:
        - name: Report Status to Port
          env:
            ACCESS_TOKEN: ${{ needs.get-access-token.outputs.access_token }}
            PERSONAL_ACCESS_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
          run: |
            status="Playlist ingestion completed successfully"
            curl --location --request POST "https://api.getport.io/v1/status" \
            --header "Authorization: Bearer $ACCESS_TOKEN" \
            --header "Content-Type: application/json" \
            --data-raw "{
                \"status\": \"$status\",
                \"description\": \"YouTube playlist ingestion job completed.\"
            }"
            echo "Reported status back to Port."
     
  ```
</details>

## Create Self Service 
We'll create a self service to trigger the workflow to run from Port

### Create blueprint 
This blueprint identifies the workflow to trigger.

<details>
   <summary><b>Self Service Blueprint (Click to expand)</b></summary>

   ```json showLineNumber
   {
     "identifier": "triggerworkflow",
     "title": "TriggerWorkflow",
     "icon": "Github",
     "schema": {
       "properties": {
         "workflow_name": {
           "icon": "Github",

>            "title": "Playlist workflow trigger",

Reviewer: Self service action doesn't need a blueprint. this is out of context

Author: Noted. 
Fixed. 
Thanks.

           "type": "string",
           "description": "Blueprint to trigger the playlist workflow"
         }
       },
       "required": ["workflow_name"]
     }
   }
   ```
</details>

### Create Action
We'll create an action that triggers the playlist workflow file

**To create the action:**

> ::info `<GITHUB_ORG_NAME>` and `<GITHUB_REPO_NAME>` with your organization and repository name.::

Reviewer: An admonition is ::: not ::
and add a title to every admonition like 
:::info Placeholders
`<GITHUB_ORG_NAME>` and `<GITHUB_REPO_NAME>` with your organization and repository name.
:::

Author: Fixed

### Create Action
We'll create an action that triggers the playlist workflow file

To create the action:
1. Open the **Self-service** tab in Port.
2. Click on **+ Action** button.
3. Click on **Edit JSON**.
4. Copy and paste the JSON below

>    :ℹ️Replace `<GITHUB_ORG_NAME>` and `<GITHUB_REPO_NAME>` with your organization and repository name.

Reviewer: bring this tip way below the points. you can leave this as a tip after all the steps and use a tip admonition

   ```json showLineNumber
   {
     "identifier": "trigger_github_workflow",
     "title": "Trigger GitHub Workflow",
     "icon": "Github",
     "description": "Triggers a GitHub workflow in the repository",
     "trigger": {
       "type": "self-service",
       "operation": "CREATE",
       "userInputs": {
         "properties": {
           "workflow_name": {
             "icon": "DefaultProperty",
             "title": "Playlist workflow action",
             "type": "string"
           }
         },
         "required": ["workflow_name"],
         "order": ["workflow_name"]
       },

>        "blueprintIdentifier": "triggerworkflow"

Reviewer: if this is the reason that you created the workflow Port will be trying to insert data into this blueprint if you use the blueprint id from port. You could change the trigger in your case so you don't need to specify blueprint else you'd have to create two SSA's

<img src="./assets/self-service.png" alt="Self-service" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/self-service-success.png" alt="service-success" style="border: 2px solid black; border-radius: 5px;">

**If Self service action is `Successful`**
Navigate to Catalog and click on Playlist service to view your playlist video data

<img src="./assets/playlist-catalog.png" alt="Playlist-catalog" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/video-entity.png" alt="Video-entity" style="border: 2px solid black; border-radius: 5px;">

## Visualization

Let's visualize our YouTube data in port by key metrics like view counts, and number of videos in the playlist.

### Setup
   - Click on catalog.
   - Click on the `+ New` button in the left sidebar.
   - Select `New dashboard`.
   - On `Title` type **Playlist Videos**
   - On `Description` type **Playlist Videos Dashboard**
   - choose an icon if desired, or leave as default
   - click Create.

<img src="./assets/Create-dashboard.jpg" alt="create-dashboard" style="border: 2px solid black; border-radius: 5px;">

**This will create a new empty dashboard.**

### Adding widgets

> 

Reviewer 3: Before introducing the widgets, add a description on what insights we're trying to gain.

Author: Done

#### Videos Count
   - Click **+ Widget** and select Number Chart.
   - Title: Videos count
   - Icon: choose an icon of your choice
   - Description: Number of videos in a playlist.
   - Chart type: Select **Count entities**
   - Blueprint: Select **Playlist**
   - Function: select **count**
   - Leave the rest as default
   - Click Save.

<img src="./assets/Videos-count-form.jpg" alt="Videos-count" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/Videos-count-dash.jpg" alt="Videos-count" style="border: 2px solid black; border-radius: 5px;">

#### Views Count
   - Click **+ Widget** and select Pie Chart.
   - Title: Views count
   - Icon: Choose an icon of your choice
   - Description: Videos view count
   - Blueprint: select **Playlist**
   - Breakdown by property: select **view count**
   - Leave the rest as default
   - Click Save.

<img src="./assets/Views-count-form.jpg" alt="Views-count" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/Views-count-dash.jpg" alt="Views-count" style="border: 2px solid black; border-radius: 5px;">

> 

Reviewer 3: Add A brief section on validating the setup:

After setting up, verify your integration by checking:
1. Data appears in Port
2. Relationships are correctly mapped
etc

# YouTube Playlist & Videos Catalog in Port

In this guide, we will demonstrate how to efficiently manage and visualize data from a YouTube playlist, providing quick insights into your video content. 
By following this guide, you'll learn how to ingest your YouTube data into Port

## Prerequisites
 - GitHub Account 
 - You have a Port account and have completed the [onboarding process](https://docs.getport.io/quickstart). 

>  - Youtube Playlist.

Reviewer 2: remove the full stop punctuations at the end of the list items

Author: Done

>  - Youtube data API V3.

Author: Added an article link

## Data Modeling
The first step is to define a model for the YouTube playlist and videos in Port. This involves creating **blueprints** for the playlist and each video, defining their properties, and establishing relationships.

### Create Blueprints

  - Click on `Builder`

>   - then click on the `+ Blueprint` button

Author: Done

           "description": "Playlist id"
         },
         "title": {
           "type": "string",
           "title": "Playlist blueprint",
           "description": "Playlist title"
         },
         "description": {
           "type": "string",
           "title": "Playlist data",
           "description": "This is playlist blueprint description"
         },
         "channel_id": {
           "type": "string",
           "title": "Channel Id",
           "description": "Channel id"
         },
         "channel_title": {
           "type": "string",
           "title": "Channel Title"
         },
         "published_at": {
           "type": "string",
           "title": "Published At",
           "description": "Published date",
           "format": "date-time"
         },
         "view_count": {
           "type": "string",
           "title": "view count"
         },
         "like_count": {
           "type": "string",
           "title": "like count"
         },
         "comment_count": {
           "type": "string",
           "title": "comment count"
         },
         "duration": {
           "type": "string",
           "title": "duration"

>          }

Author: YouTube data API has them modelled has strings

       },
       "required": []
     },
     "mirrorProperties": {},
     "calculationProperties": {},
     "aggregationProperties": {},
     "relations": {}
   }
        
  ```
</details>

<img src="./assets/playlist-blueprint.png" alt="Playlist-blueprint" style="border: 2px solid black; border-radius: 5px;">

<details>
  <summary><b> Video Blueprint JSON (Click to expand)</b></summary>

  ```json showLineNumber
  {
    "identifier": "video",
    "description": "YouTube videos in a playlist",
    "title": "Video",
    "icon": "Youtrack",
    "schema": {
      "properties": {
        "id": {
          "icon": "Youtrack",
          "type": "string",
          "description": "Video id",
          "title": "Id"

>         },

Author: Id and title removed

        "channel_id": {
          "icon": "DefaultProperty",
          "type": "string",
          "title": "Channel ID",
          "description": "Playlist channel id"
        },
        "channel_title": {
          "type": "string",
          "title": "Channel Title",
          "description": "Playlist channel title"
        },
        "category_id": {
          "type": "string",
          "title": "Category Id"
        },
        "duration": {
          "type": "string",
          "title": "Duration"
        },
        "caption": {
          "type": "string",
          "title": "Caption"
        },
        "view_count": {
          "type": "string",
          "title": "view count"
        },
        "like_count": {
          "type": "string",
          "title": "like count"
        },
        "comment_count": {
          "type": "string",
          "title": "comment count"
        }
      },
      "required": []
    },
    "mirrorProperties": {},
    "calculationProperties": {},
    "aggregationProperties": {},
    "relations": {
      "playlist": {
        "title": "Playlist",
        "description": "Playlist",
        "target": "playlist",
        "required": false,

>         "many": false

Author: I don't understand this

                    echo "Uploaded video $title ($video_id) to Port."
                done

                # Check if there is another page of results
                nextPageToken=$(echo "$response" | jq -r '.nextPageToken')
                [ "$nextPageToken" == "null" ] && break
            done

    report-status:
      runs-on: ubuntu-latest
      needs: upsert-playlist-data
      steps:
        - name: Report Status to Port
          env:
            ACCESS_TOKEN: ${{ needs.get-access-token.outputs.access_token }}
            PERSONAL_ACCESS_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
          run: |
            status="Playlist ingestion completed successfully"
            curl --location --request POST "https://api.getport.io/v1/status" \
            --header "Authorization: Bearer $ACCESS_TOKEN" \
            --header "Content-Type: application/json" \
            --data-raw "{
                \"status\": \"$status\",
                \"description\": \"YouTube playlist ingestion job completed.\"
            }"
            echo "Reported status back to Port."
     
  ```
</details>

## Create Self Service 
We'll create a self service to trigger the workflow to

> run from port

Author: Fixed

### Create blueprint 
This blueprint identifies the workflow to trigger.

<details>
   <summary><b>Self Service Blueprint (Click to expand)</b></summary>

   ```json showLineNumber
   {
     "identifier": "triggerworkflow",
     "title": "TriggerWorkflow",
     "icon": "Github",
     "schema": {
       "properties": {
         "workflow_name": {
           "icon": "Github",
           "title": "Playlist workflow trigger",
           "type": "string",
           "description": "Blueprint to trigger the playlist workflow"
         }
       },
       "required": ["workflow_name"]
     }
   }

>    ```

Author: Created a self service to trigger the workflow. To create a self service, I needed to create a self service blueprint

- **`reportWorkflowStatus`** allows Port to report the workflow status back to the Port dashboard.

### **Trigger the Action**

To execute the action:
  1. Open the **Self-service** tab in Port.
  2. Select **Trigger GitHub Workflow**.
  3. Enter the name of the workflow file (e.g., `run-workflow.yml`) as the `workflow_name`.
  4. Submit the form, and the GitHub workflow will run automatically.

You can view the status of the action by clicking on the progress icon at the top right corner on the page
As shown on the images below 

<img src="./assets/self-service.png" alt="Self-service" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/self-service-success.png" alt="service-success" style="border: 2px solid black; border-radius: 5px;">

**If Self service action is `Successful`**
Navigate to Catalog and click on Playlist service to view your playlist video data

<img src="./assets/playlist-catalog.png" alt="Playlist-catalog" style="border: 2px solid black; border-radius: 5px;">

<img src="./assets/video-entity.png" alt="Video-entity" style="border: 2px solid black; border-radius: 5px;">

## Visualization

> Let's visualize our YouTube data in port by key metrics like view counts, and number of videos in the playlist.

Author: Fixed
